bread/contrib/triggers/models.py

```diff
@@ -0,0 +1,170 @@
+            email = email.strip()
+            if email.startswith("@"):
+                user = get_user_model().objects.filter(username=email[1:]).first()
+                if user and user.email:
+                    recepients.append(user.email)
+                else:
+                    users = Group.objects.filter(name=email[1:]).first().user_set.all()
+                    recepients.extend(u.email for u in users if u.email)
+            elif is_email_simple(email):
+                recepients.append(email)
+            else:  # try to get value from object via accessor
+                extracted_email = hg.resolve_lookup(object, email) or ""
+                if is_email_simple(extracted_email):
+                    recepients.append(extracted_email)
+        if recepients:
+            send_mail(
+                subject=engines["django"]
+                .from_string(self.subject)
+                .render({"object": object}),
+                message=engines["django"]
+                .from_string(self.message)
+                .render({"object": object}),
+                from_email=None,
+                recipient_list=recepients,
+            )
+
+    class Meta:
+        verbose_name = _("Send Email Action")
+        verbose_name_plural = _("Send Email Actions")
+
+
+class Trigger(models.Model):
+    model = models.ForeignKey(
+        ContentType,
+        verbose_name=_("Model"),
+        on_delete=models.PROTECT,
+    )
+    filter = QuerysetField(_("Filter"), modelfieldname="model")
+    enable = models.BooleanField(default=True)
+    action = models.ForeignKey(Action, on_delete=models.PROTECT)
+
+    class Meta:
+        abstract = True
+
+
+class DataChangeTrigger(Trigger):
+    type = models.TextField(
+        _("Type"),
+        choices=(
+            ("added", _("Added")),
+            ("changed", _("Changed")),
+            ("deleted", _("Deleted")),
+        ),
+    )
+
+    def __str__(self):
+        return f"{self.model} Trigger: {self.action} when {self.type})"
+
+    class Meta:
+        verbose_name = _("Data change trigger")
+        verbose_name_plural = _("Data change triggers")
+
+
+INTERVAL_CHOICES = {
+    "minutes": (datetime.timedelta(minutes=1), _("Minutes")),
+    "hours": (datetime.timedelta(hours=1), _("Hours")),
+    "days": (datetime.timedelta(days=1), _("Days")),
+    "weeks": (datetime.timedelta(days=7), _("Weeks")),
+    "months": (datetime.timedelta(days=30.5), _("Months")),  # approximation
+    "years": (datetime.timedelta(days=365.25), _("Years")),  # approximation
+}
+
+
+class DateFieldTrigger(Trigger):
+    field = models.CharField(max_length=255)
+    offset_type = models.CharField(
+        max_length=255,
+        choices=tuple((name, value[1]) for name, value in INTERVAL_CHOICES.items()),
+    )
+    offset_amount = models.IntegerField(
+        help_text=_("Can be negative (before) or positive (after)")
+    )
+
+    def triggerdate(self, object):
```

Comment:
I would appreciate a typehint for what exactly this method returns

Reply:
Added, although I was thinking the name would be clear enough.

Reply:
the name `triggerdatetime` would have been clear enough :)

Reply:
Well, I would prefer to *not* abuse names for type-hinting ;)